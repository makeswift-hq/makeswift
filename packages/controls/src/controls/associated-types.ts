import { ControlDefinition } from './definition'

export type ControlType<D> =
  D extends ControlDefinition<infer ControlType, any, any, any, any, any>
    ? ControlType
    : never

export type ConfigType<D> =
  D extends ControlDefinition<any, infer Config, any, any, any, any>
    ? Config
    : never

export type DataType<D> =
  D extends ControlDefinition<any, any, infer DataType, any, any, any>
    ? DataType
    : never

export type ValueType<D> =
  D extends ControlDefinition<any, any, any, infer ValueType, any, any>
    ? ValueType
    : never

export type ResolvedValueType<D> =
  D extends ControlDefinition<any, any, any, any, infer ResolvedValueType, any>
    ? ResolvedValueType
    : never

export type InstanceType<D> =
  D extends ControlDefinition<any, any, any, any, any, infer InstanceType>
    ? InstanceType
    : never
