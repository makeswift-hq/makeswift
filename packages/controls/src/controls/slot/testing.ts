import { SlotDefinition } from './slot'

export const renderedNode = Symbol('rendered node')

export type RenderedNode = typeof renderedNode

class Definition extends SlotDefinition<RenderedNode> {}

export function Slot(): Definition {
  return new Definition()
}

export { Definition as SlotDefinition }
