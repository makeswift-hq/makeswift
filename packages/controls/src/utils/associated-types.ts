export type WithAssociatedTypes<T> = {
  readonly __associated_types__?: T
}

export type AssociatedTypes<T> = T extends {
  __associated_types__?: infer Types
}
  ? Types extends Record<string, unknown>
    ? Types
    : never
  : never

export type HasAssociatedTypes<T> = T extends {
  __associated_types__?: infer Types
}
  ? Types extends Record<string, unknown>
    ? true
    : false
  : false

export type AssociatedType<
  T,
  K extends keyof AssociatedTypes<T>,
> = AssociatedTypes<T>[K]
