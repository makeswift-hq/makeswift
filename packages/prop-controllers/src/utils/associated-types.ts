export type WithAssociatedTypes<T> = {
  readonly __associated_types__?: T
}

export type AssociatedTypes<T> = T extends {
  __associated_types__?: infer Types
}
  ? Types extends Record<string, unknown>
    ? Types
    : never
  : never

export type AssociatedType<
  T,
  K extends keyof AssociatedTypes<T>,
> = AssociatedTypes<T>[K]
