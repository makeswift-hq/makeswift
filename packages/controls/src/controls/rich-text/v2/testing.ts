import { z } from 'zod'

import { type SerializedRecord } from '../../../serialization'
import { type DataType } from '../../associated-types'
import { serialize } from '../../definition'
import {
  ControlInstance,
  DefaultControlInstance,
  type SendMessage,
} from '../../instance'

import { RichTextPluginControl } from './plugin'
import { RichTextDefinition } from './rich-text'

export const renderedNode = Symbol('rendered node')

export type RenderedNode = typeof renderedNode

type UserConfig = z.infer<typeof Definition.schema.userConfig>

class Definition extends RichTextDefinition<RenderedNode> {
  createInstance(sendMessage: SendMessage<any>): ControlInstance<any> {
    return new DefaultControlInstance(sendMessage)
  }

  serialize(): [SerializedRecord, Transferable[]] {
    return serialize(this.config, {
      type: Definition.type,
    })
  }

  toText(_data: DataType<RichTextDefinition<string>>): string {
    return ''
  }

  get pluginControls(): RichTextPluginControl[] {
    return []
  }

  pluginControlAt(_index: number) {
    return undefined
  }
}

export { Definition as RichTextDefinition }

export const RichText = (config?: UserConfig) => new Definition(config ?? {})

RichText.Mode = Definition.Mode
