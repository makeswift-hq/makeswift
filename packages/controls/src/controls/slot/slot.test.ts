import { type ResolvedValueType } from '../associated-types'

import { renderedNode, Slot, SlotDefinition } from './testing'

describe('Slot', () => {
  describe('constructor', () => {
    test('resolved value type', () => {
      const slot = Slot()
      const node: ResolvedValueType<typeof slot> = renderedNode

      expect(slot).toMatchSnapshot()
      expect(node).toBe(renderedNode)
    })

    test('disallows extraneous properties', () => {
      // @ts-expect-error
      Slot({
        extra: 'extra',
      })
    })
  })

  describe('assignability', () => {
    function assignTest(_def: SlotDefinition) {}
    assignTest(Slot())
  })
})
