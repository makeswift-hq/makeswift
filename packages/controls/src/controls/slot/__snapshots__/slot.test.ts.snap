// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Slot constructor resolved value type 1`] = `
Definition {
  "config": {},
}
`;
